src/sentry/api/endpoints/project_rule_actions.py

```diff
@@ -60,7 +65,66 @@ def post(self, request: Request, project) -> Response:
             project, platform=project.platform, default="javascript", tagged=True
         )
 
+        if features.has(
+            "projects:verbose-test-alert-reporting", project=project, user=request.user
+        ):
+            return self.execute_future_on_test_event(test_event, rule)
+        else:
+            # Old existing behavior to wrap this in a handler which buries exceptions
+            for callback, futures in activate_downstream_actions(rule, test_event).values():
+                safe_execute(callback, test_event, futures)
+            return Response()
+
+    def execute_future_on_test_event(
+        self,
+        test_event: Event | None,
+        rule: Rule,
+    ) -> Response:
+        """
+        A slightly modified version of utils.safe.safe_execute that handles
+        IntegrationFormErrors, and returns a body with `{ actions: [<error info>] }`.
+
+        This is used in our Alert Rule UI to display errors to the user.
+        """
+        action_exceptions = []
         for callback, futures in activate_downstream_actions(rule, test_event).values():
-            safe_execute(callback, test_event, futures)
+            try:
+                callback(test_event, futures)
+            except Exception as exc:
+                if hasattr(callback, "im_class"):
+                    cls = callback.im_class
+                else:
+                    cls = callback.__class__
+                callback_name = getattr(callback, "__name__", str(callback))
+                cls_name = cls.__name__
+                logger = logging.getLogger(f"sentry.test_rule.{cls_name.lower()}")
+
+                # safe_execute logs these as exceptions, which can result in
+                # noisy sentry issues, so log with a warning instead.
+                if isinstance(exc, IntegrationFormError):
+                    logger.warning(
+                        "%s.test_alert.integration_form_error",
+                        callback_name,
+                        extra={"exception": exc},
+                    )
+
+                    # IntegrationFormErrors should be safe to propagate via the API
+                    action_exceptions.append(str(exc))
```

Comment:
Do we want to expose `exc.field_errors`? The GroupIntegrationDetails endpoint exposes this property in API responses, and doing the same here could help with consistent error responses from integrations.

Reply:
That's true, but I'd need to update the UI to handle this whereas this pattern is already supported. I'm also debating whether we should be exposing general `IntegrationError`  instead, but I'm not sure how safe it would be to do this.